### ISSUE TITLE
OnDetach will not be called on page reload

### ISSUE THREAD

**Issue by Maurice-Mueller** [vaadin/flow#13221]
### Description of the bug

The [documentation](https://vaadin.com/docs/latest/flow/components/creating-components/lifecycle-callbacks) states 

> The onAttach() method is invoked when the Component has attached to the UI. Its counterpart, the onDetach() method, is invoked right before the component detaches from the UI. **These are good points at which to reserve and release resources used by the component.**

This is what I did. Sadly, the onDetach() method is not called if a user decides to reload the page. Thus, all resources acquired during onAttach() leak.

### Expected behavior

onDetach() will always be called if the component is detached - regardless the reason.

### Minimal reproducible example

public class TestClass extends Component {

   @Override
   protected void onAttach(AttachEvent attachEvent) {
      super.onAttach(attachEvent);
      System.out.println("Calling onAttach");
   }

   @Override
   protected void onDetach(DetachEvent detachEvent) {
      super.onDetach(detachEvent);
      System.out.println("Calling onDetach");
   }
}

### Versions

- Vaadin / Flow version:
- Java version: kotlin with JDK 17
- OS version: Arch Linux (current)
- Browser version (if applicable): Firefox 97

**Comment by Artur-** [comment]
onDetach is indeed called when the component  is detached from the UI. The problem is that the UI is not detached from the VaadinSession when you refresh in the browser but only later on

**Comment by Maurice-Mueller** [comment]
Thanks for the reply! 

On a refresh the onAttach method is called again. So you basically have this behavior:

1. user navigates to a page
2. onAttach is called
3. user refreshes page
4. onAttach is called
5. user leaves page
6. onDetach is called

So onAttach is called twice and onDetach only called once. Therefore the resource leak.

**Comment by tarekoraby** [comment]
Related to #6293.

**Comment by tarekoraby** [comment]
There is a workaround based on Beacon API documented at https://cookbook.vaadin.com/notice-closed.

**Comment by Maurice-Mueller** [comment]
@taefi I don't think this is an enhancement but a bug. Sure, you need to add to the code but as far as I understand the API contract: for every onAttach() call there should be a corresponding onDettach() call. Otherwise the statement in the docs does not make any sense ("These are good points at which to reserve and release resources used by the component.")
So the API contract is broken and needs to be fixed.

**Comment by Artur-** [comment]
To be clear, `onDetach` is always called and the API contract is not broken. It is called way later than you would expect though (when the session expires).

No matter if we call this enhancement or bug or anything else, I think it should be improved (fixed). For instance by integrating the mentioned cookbook recipe.

**Comment by Maurice-Mueller** [comment]
Ah ok, I didn't get the part with the expired session. Thanks for clarification!

**Comment by mshabarov** [comment]
@Maurice-Mueller this has been fixed in Vaadin 24.1.0.alpha4 by implementing https://github.com/vaadin/flow/issues/6293. Now when you close a tab in browser or refresh a page, `onDetach` listener is called immediately. I close this ticket.
